keps/sig-node/1977-container-notifier/README.md

```diff
@@ -0,0 +1,770 @@
+        // +optional
+        PodNotificationStatuses []PodNotificationStatus
+}
+
+// PodNotificationStatus represents the current state of a Notification for a specific pod
+type PodNotificationStatus struct {
+        // This field is required
+        PodName string
+
+        // This field is required
+        ContainerNotificationStatuses []ContainerNotificationStatus
+ }
+
+// ContainerNotificationStatus represents the current state of a Notification for a specific container
+// The "Succeed" parameter is false until it completes successfully when it will be changed to true.
+// If an error occurs, the "Error" parameter will be set.
+type ContainerNotificationStatus struct {
+        // This field is required
+        ContainerName string
+
+         // If not set, indicating Action has not started
+         // If set, it means Action has started at the specified time
+        // +optional
+        StartTimestamp *int64
+
+        // Succeed is set to true when the action is executed in the container successfully.
+        // It will be set to false if the action cannot be executed successfully after  
+        // TimeoutSeconds in ContainerNotifierAction passes or there is any error.
+        // +optional
+        Succeed *bool
+
+        // The last error encountered when executing the action.
+        // The container notifier controller (or Kubelet in phase 2) 
+        // might update this field each time it retries the execution.
+        // +optional
+        Error *ActionError
+}
+
+type ActionError struct {
+        // Type of the error
+        // This is required
+        ErrorType ErrorType
+
+        // Error message
+        // +optional
+        Message *string
+
+        // More detailed reason why error happens
+        // +optional
+        Reason *string
+	
+        // It indicates when the error occurred
+        // +optional
+        Timestamp *int64
+}
+
+type ErrorType string
+
+// More error types could be added, e.g., Forbidden, Unauthorized, etc.
+const (
+        // The Notification times out
+        Timeout ErrorType = "Timeout"
+
+        // The Notification fails with an error
+        Error ErrorType = "Error"
+)
+```
+
+External controller deletes the Notification API objects after it is done.
+
+### Phase 2 API Additions
+
+In Phase 2, add a "signal" field to ContainerNotifierHandler.
+
+```
+// Handler defines a specific action that should be taken or a signal that should be delivered
+// Only one of Exec and Signal should be set and not both
+type ContainerNotifierHandler struct {
+    // Exec specifies the action to take.
+    // +optional
+    Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
+
+    // Signal specifies a signal to send to the container
+    // +optional
+    // define constants for signals?
+    // validate the signals are valid?  windows?
+++    Signal string
+}
+```
+
+### Phase 3 API Additions
```

Comment:
I think we should say something like "Phase 3 is more loosely decided than phases 1 and 2, and if it is needed, we will have to revisit the design details".  There are just too many variables to see the needs at this point (IMO).

As such, I am not aggressively API reviewing it :)

Reply:
Ok